### PR TITLE
Node controller supports disabling node probes.

### LINKED ISSUE #4565: Node controller to support disabling node probes.
Split off #156 as a smaller more specific work item.

Node controller should have a flag that makes it not attempt to GET /healthz from kubelets.

This would be used in conjunction with #4562.  So, it would rely on kubelets to post the node status.  However, it would still need to check periodically check each node to see if it `nodeStatus.condition[].lastProbeTime` had been updated recently.  If it hadn't, the node controller would still kick the pods off the node.

After we have converted the top N Getting-started-guides to the new style, we can rip out the kubelet client from the node controller altogether.

### PR DESCRIPTION
**Pull request by jszczepkowski**
Node controller supports disabling sending node probes and updating node statuses. Controlled by --sync_node_status flag. Resolves #4565.

### CODE REVIEW

**Comment by jszczepkowski** [comment]
@ddysher please take a look as this is related to #4562.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 1]
```diff
@@ -32,12 +32,15 @@ import (
...
 	"github.com/golang/glog"
+	flag "github.com/spf13/pflag"
 )
 
 var (
 	ErrRegistration   = errors.New("unable to register all nodes.")
 	ErrQueryIPAddress = errors.New("unable to query IP address.")
 	ErrCloudInstance  = errors.New("cloud provider doesn't support instances.")
+
+	syncNodeStatus = flag.Bool("sync_node_status", true, "Should node controler send probes to kubelets and update NodeStatus.")
```
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: All flags are defined in hyber-kube now, you should define it there.   see cmd/kube-controller-manager.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 2]
```diff
@@ -216,6 +227,33 @@ func (s *NodeController) SyncNodeStatus() error {
...
+	}
+	return nil
+}
+
+func latestReadyTime(node *api.Node) util.Time {
+	time := node.ObjectMeta.CreationTimestamp
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == api.NodeReady &&
+			condition.Status == api.ConditionFull &&
+			condition.LastProbeTime.After(time.Time) {
```
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: what is condition.LastProbeTime.After(time.Time)?

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 3]
```diff
@@ -114,12 +116,21 @@ func (s *NodeController) Run(period time.Duration, syncNodeList bool) {
...
+		// Start syncing node status.
+		go util.Forever(func() {
+			if err = s.SyncNodeStatus(); err != nil {
+				glog.Errorf("Error syncing status: %v", err)
+			}
+		}, period)
+	} else {
+		// Start checking node reachability.
+		go util.Forever(func() {
+			if err = s.CheckNodeReachable(); err != nil {
```
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: CheckNodeReachable is confusing.  It's not check node reachable, it's doing pod eviction.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 2 above)
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: Oh, I see.  You assign CreationTimestamp to time.... can we s/time/creationTime/?  overriding package name sounds weired.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 1 above)
Comment (jszczepkowski) [inline pkg/cloudprovider/controller/nodecontroller.go]: Done. I've added the flag to controlmanager.go, the same way as it works for sync_nodes flag.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 2 above)
Comment (jszczepkowski) [inline pkg/cloudprovider/controller/nodecontroller.go]: Done.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 3 above)
Comment (jszczepkowski) [inline pkg/cloudprovider/controller/nodecontroller.go]: Done.

**Comment by ddysher** [comment]
LGTM.  Thanks!

**Comment by jszczepkowski** [comment]
Commits squashed.

**Comment by ddysher** [comment]
Thanks.  FYI.  I expect the code to be refactored later once POST node/pod status is in place.

**Pull request merged**